### PR TITLE
Extension rewrite

### LINKED ISSUE #27: An error is raised when generating a png
**What happened**:
Generating a png for a workflow seems to throw an error under certain conditions:
image

**What you expected to happen**:
The png to be generated.

**How to reproduce it**:
- Create a workflow.json file containing:
```
{
  "id": "helloworld",
  "version": "1.0.0",
  "specVersion": "0.8",
  "name": "Hello World Workflow",
  "description": "Inject Hello World",
  "start": "hello state",
  "states": [
     {
        "name": "hello state",
        "type": "inject",
        "data": {
           "result": "Hello World!"
        },
        "end": true
     }
  ]
}
```
- ctrl+shift+p > Export Diagram to PNG (or Generate Diagram, will also fail)

**Anything else we need to know?**:
[From Saad Ansari on Slack ](https://cloud-native.slack.com/team/U0676HP4Q85)

It seems to be related to a incorrect handling of paths and/or expected inputs of the command.

The extension also relies on external services to generate the files which is not necessary as VSCode already packs browser's capatiblities to render the diagram.

Finally, the code base is scattered and difficult to debug.

**Environment**:
Tested on Win11 x64. Reported from MacOS.

### PR DESCRIPTION
**Pull request by JBBianchi**
**Many thanks for submitting your Pull Request :heart:!**

**What this PR does / why we need it**:
This is a complete rewrite of the extension. It simplifies & cleans the code and remove the dependency to an external server to render the graph.

**Special notes for reviewers**:
Fix #27 

**Additional information (if needed):**

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -1 +1 @@
-* @tsurdilo @manuelstein @ricardozanini
\ No newline at end of file
+* @tsurdilo @manuelstein @ricardozanini @JBBianchi
```
Comment (ricardozanini) [inline .github/CODEOWNERS:1]: You can remove Manuel here and I'll give you write permissions here. :)

**Review by ricardozanini — APPROVED** [review]
Looks neat! I'd like to see this in action now. Is there a way to install a snapshot?

**Comment by JBBianchi** [comment]
> Looks neat! I'd like to see this in action now. Is there a way to install a snapshot?

I think you should be able to test it out using the [Building from source](https://github.com/serverlessworkflow/vscode-extension?tab=readme-ov-file#building-from-source) instructions. The output file (vsix) will have a different name though, `serverless-workflow-0.0.1.vsix` if I'm not mistaken.

**Comment by ricardozanini** [comment]
@JBBianchi  I was able to export to SVG and preview, but for some reason, I couldn't for PNG. The preview pane just opens and shows the diagram when I pick this option. Is this expected?

**Comment by JBBianchi** [comment]
> @JBBianchi I was able to export to SVG and preview, but for some reason, I couldn't for PNG. The preview pane just opens and shows the diagram when I pick this option. Is this expected?

It's not, I'll investigate. Did the workflow you were trying to export exist on the disk or was it just a "virtual" file (tab) in VSCode but not yet saved on disk ?

**Comment by JBBianchi** [comment]
I pushed a fix for "virtual files" but I think I faced a race condition on the first time I tried to debug but couldn't reproduce it.

@ricardozanini can you give me some feedbacks using the latest committed version ?

**Comment by ricardozanini** [comment]
> > @JBBianchi I was able to export to SVG and preview, but for some reason, I couldn't for PNG. The preview pane just opens and shows the diagram when I pick this option. Is this expected?
> 
> It's not, I'll investigate. Did the workflow you were trying to export exist on the disk or was it just a "virtual" file (tab) in VSCode but not yet saved on disk ?

It exists! I'll try the new commit.

**Comment by ricardozanini** [comment]
@JBBianchi same thing, the preview even flashes a little bit now, but won't generate the png. We can do a live session tomorrow, wdyt?

**Comment by JBBianchi** [comment]
Sure, with pleasure.

**Comment by JBBianchi** [comment]
I found a bug when generating the PNG. The process to generate the PNG is the following:
- Generate the SVG using Mermaid
- Grab the SVG, convert it to base64
- Use the base64 as an img tag source
- When the image has loaded, inject it in a canvas and export the canvas as PNG

There was a problem generating the base64 from the SVG in some cases (I guess bad character escaping). It should be fixed now. (If GitHub would be kind enough to sync the PR ...)

**Comment by ricardozanini** [comment]
I'll retest again this afternoon! Thanks!

**Comment by ricardozanini** [comment]
@JBBianchi still not working on my machine :(

**Comment by ricardozanini** [comment]
Hey @JBBianchi I asked a friend (@handreyrc) to also test the extension and he used this file:

```json
{
  "id": "statesExample",
  "version": "1.0",
  "specVersion": "0.8",
  "name": "States Example",
  "description": "States Example",
  "start": "Create Initial Data",
  "states": [
   {
    "name": "Create Initial Data",
    "type": "sleep",
    "end": false,
    "transition": "Inject data"
   },
   {
    "name": "Inject data",
    "type": "foreach",
    "end": false,
    "compensatedBy": "Compensate",
    "transition": "Wait for Events",
    "onErrors": [
     {
      "errorRef": "SomeError1",
      "end": false,
      "transition": "Handle Error1"
     },
     {
      "errorRef": "SomeError2",
      "end": false,
      "transition": "Handle Error2"
     }
    ]
   },
   {
    "name": "Handle Error1",
    "type": "inject",
    "end": true
   },
   {
    "name": "Handle Error2",
    "type": "inject",
    "end": true
   },
   {
    "name": "Wait for Events",
    "type": "event",
    "onEvents": [
     {
      "eventRefs": [
       "Event"
      ],
      "actions": [
       {
        "functionRef": "Function"
       }
      ]
     }
    ],
    "transition": "Run Operations",
    "end": false
   },
   {
    "name": "Run Operations",
    "type": "operation",
    "actionMode": "sequential",
    "actions": [
     {
      "name": "action1",
      "functionRef": "Task1"
     },
     {
      "name": "action2",
      "subFlowRef": "Task2"
     }
    ],
    "end": false,
    "transition": "Switch State"
   },
   {
    "name": "Switch State",
    "type": "switch",
    "onErrors": [
     {
      "errorRef": "SomeError3",
      "end": false,
      "transition": "Handle Error3"
     }
    ],
    "eventConditions": [
     {
      "eventRef": "event1",
      "transition": "Event1 State"
     },
     {
      "eventRef": "event2",
      "transition": "Event2 State"
     }
    ],
    "defaultCondition": {
     "transition": "Default State"
    }
   },
   {
    "name": "Handle Error3",
    "type": "callback",
    "end": true
   },
   {
    "name": "Event1 State",
    "type": "event",
    "end": true
   },
   {
    "name": "Event2 State",
    "type": "event",
    "end": true
   },
   {
    "name": "Default State",
    "type": "parallel",
    "end": true
   },
   {
    "name": "Compensate",
    "type": "callback",
    "end": true
   }
  ]
 }
```

Here's his output:

> No the PNG export does not work, it opens a new preview tab for me and no png file is created.
I tested the SVG export as well, it seems it quickly creates a new tab and then closes it in the process. The SVG file was created but I cannot open it, it is broken.

**Comment by JBBianchi** [comment]
Afaik, he is not able to preview it either because the TS SDK raises an error: `Format not supported`

**Comment by handreyrc** [comment]
> Afaik, he is not able to preview it either because the TS SDK raises an error: `Format not supported`

Hey @JBBianchi,

@ricardozanini posted a workflow not supported by this extension I mentioned in a talk with him. You are right this workflow won't work.
I apologize for the misunterstood!

In fact I tested this PR with many other worflows and it does not work.  
Just giving you more background about the behavior above:
OS: Fedora Linux 38
VsCode version: 1.87.0  

Here goes one of the workflows it fails:
[greeting.sw.json](https://github.com/serverlessworkflow/vscode-extension/files/14526324/greeting.sw.json)

This is how it looks on my machine. I hope it helps you.
Extension_Test

**Comment by JBBianchi** [comment]
> > Afaik, he is not able to preview it either because the TS SDK raises an error: `Format not supported`
> 
> Hey @JBBianchi,
> 
> @ricardozanini posted a workflow not supported by this extension I mentioned in a talk with him. You are right this workflow won't work. I apologize for the misunterstood!
> 
> In fact I tested this PR with many other worflows and it does not work. Just giving you more background about the behavior above: OS: Fedora Linux 38 VsCode version: 1.87.0
> 
> Here goes one of the workflows it fails: [greeting.sw.json](https://github.com/serverlessworkflow/vscode-extension/files/14526324/greeting.sw.json)
> 
> This is how it looks on my machine. I hope it helps you. ...

Thanks for your report. I'll investigate.

**Comment by JBBianchi** [comment]
@handreyrc I pushed an update that should provide more debug info. Could you please try again and tell me what error pops up ?

Two more little things that could bring some pieces of info:
- switch to the explorer view
- enable the Developer Tools console to see if some messages are displayed there (F1: Developer: Toggle Developer Tools)
image

**Comment by JBBianchi** [comment]
I fixed a possible race condition where the request to draw the graph would be sent before mermaid was important in the viewport (basically, the handler for the rendering request wouldn't be attached yet, therefore resulting in a blank page).

Besides that, which didn't seem to be the case in @handreyrc case as the graph was rendering, I didn't find anything else and it seems to be working on the latest Fedora:
fedora

**Comment by ricardozanini** [comment]
@JBBianchi @handreyrc it's working now!!!

**Review by ricardozanini — APPROVED** [review]

**Comment by ricardozanini** [comment]
@cdavernas @antmendoza can you please review?

**Review by cdavernas — APPROVED** [review]
LGTM! Cheers ❤️

**Comment by ricardozanini** [comment]
@antmendoza can you ptal?

**Comment by LuboTerifaj** [comment]
Hello @JBBianchi , @ricardozanini 

is there any issue reported with the details what is its goal? I see only one bug-fix linked in the PR description. However, the PR's description looks like it has a wider scope than just to fix a bug.

I see there is no github action that would build the extension. Building it locally is the only option at the moment?

Thanks!

**Comment by ricardozanini** [comment]
@LuboTerifaj there's none AFAIK. The extension won't work in the current context. @JBBianchi did the whole rewriting to make it work again. So what we are trying to do here is to test it locally and see if it's working. e.g. visualize the workflow and export to PNG/SVG.

@cdavernas Wanna take a look? After your approval, we can consider merging it. @JBBianchi can we follow up by fixing also the web editor? 🤔

**Comment by ricardozanini** [comment]
@LuboTerifaj one more thing, we don't have a GHA for this extension. If you know about GHA, can you consider a PR? 🙏 
thank you!

**Comment by cdavernas** [comment]
@ricardozanini I have already approved it yesterday 😉

**Comment by JBBianchi** [comment]
>Hello @JBBianchi , @ricardozanini
>
>is there any issue reported with the details what is its goal? I see only one bug-fix linked in the PR description. However, the PR's description looks like it has a wider scope than just to fix a bug.
>
>I see there is no github action that would build the extension. Building it locally is the only option at the moment?
>
>Thanks!

@LuboTerifaj > The first goal was to fix a bug reported on Slack. But when I tracked the bug I noticed 1st that the extension was relying on an external server to process the graph, which isn't optimal, and 2nd the code base was a bit scattered and not very ubiquitous. Therefore I decided to go for a full refactor, bringing the graph generation inside the extension.

At the moment, we didn't configure any pipeline to build the extension so there only solution to test it is to clone it and build it locally. But it's certainly possible to create a pipeline to do so. Please feel free to contribute one if you feel like it.


> @LuboTerifaj there's none AFAIK. The extension won't work in the current context. @JBBianchi did the whole rewriting to make it work again. So what we are trying to do here is to test it locally and see if it's working. e.g. visualize the workflow and export to PNG/SVG.
> 
> @cdavernas Wanna take a look? After your approval, we can consider merging it. @JBBianchi can we follow up by fixing also the web editor? 🤔

What fix for the web editor are you talking about @ricardozanini ? (Besides updating the TS SDK. It's not related to this VSCode extension.)

Other general note, there is no tests for the extension. I didn't dig a lot but at first glance I couldn't find a proper e2e testing example for VSCode.

**Comment by ricardozanini** [comment]
> @ricardozanini I have already approved it yesterday 😉

LOL

**Comment by ricardozanini** [comment]
@JBBianchi I thought the web editor in the website would share the same libs/framework as the extension. My bad then! Can we merge this?

**Comment by JBBianchi** [comment]
> @JBBianchi I thought the web editor in the website would share the same libs/framework as the extension. My bad then! Can we merge this?

They do, the SDK and Mermaid :) We can merge.

**Comment by LuboTerifaj** [comment]
Just fyi, the user experience is not very good when you try to open more workflows in a short time. It is very time consuming to open the diagram using Palette again and again. I have reported the issue here: https://github.com/serverlessworkflow/vscode-extension/issues/29

**Comment by ricardozanini** [comment]
Thank you @LuboTerifaj!

**Pull request merged**